### PR TITLE
Small policy calculation optimization

### PR DESCRIPTION
**Pull request by farmersrice**
3-4% speedup in random backend. May be noise or hardware dependent, it would be appreciated if more people could test it.

```
./lc0_policy_optimized benchmark --backend=random
===========================
Total time (ms) : 340446
Nodes searched  : 69693334
Nodes/second    : 204711

./lc0_c4c92f6 benchmark --backend=random
===========================
Total time (ms) : 340487
Nodes searched  : 67522311
Nodes/second    : 198310

./lc0_policy_optimized benchmark (t59)
===========================
Total time (ms) : 340808
Nodes searched  : 42638750
Nodes/second    : 125110

./lc0_c4c92f6 benchmark (t59)
===========================
Total time (ms) : 340724
Nodes searched  : 42673511
Nodes/second    : 125243
```

### CODE REVIEW

**Comment by farmersrice** [comment]
from Misuzu, Windows 7, i7-3930k:
```
.\lc0.exe benchmark --backend=random --threads=2
c4c92f6  Nodes/second   : 130631
PR1278  Nodes/second    : 134029
```

**Comment by Naphthalin** [comment]
Comparing to https://github.com/LeelaChessZero/lc0/pull/1150#issuecomment-620771542:
```
lc0_1278 benchmark --backend=random

Total time (ms) : 340341
Nodes searched  : 43478940
Nodes/second    : 127751

lc0_1278 benchmark -w weights\591215

Total time (ms) : 340296
Nodes searched  : 17219505
Nodes/second    : 50601
```

So I see with `backend=random` 2.5% nps increase compared to the AppVeyor build (1.6% to official build), and 0.8% with `591215` (same as official build). Good find I'd say -- I hope it doesn't introduce any unwanted behavior with parallelism or so :)

**Comment by borg323** [comment]
Note the comment that is being deleted: `Edge::SetP does some rounding, so only add to the total after rounding.` Maybe this is not an issue after #912 but it needs careful consideration.

File: src/mcts/search.cc [hunk 1]
```diff
@@ -1444,16 +1445,14 @@ void SearchWorker::FetchSingleNodeResult(NodeToProcess* node_to_process,
...
-    // format. Normalization happens later.
-    edge.edge()->SetP(p);
-    // Edge::SetP does some rounding, so only add to the total after rounding.
-    total += edge.edge()->GetP();
+    intermediate_[counter++] = p;
+    total += p;
   }
+  counter = 0;
   // Normalize P values to add up to 1.0.
   if (total > 0.0f) {
```
Comment (Tilps) [inline src/mcts/search.cc:1453]: This code path is pretty much always hit (it takes some truly crazy corner case to not be true IIRC), but if it isn't, edge SetP will never be called now.

**Comment by farmersrice** [comment]
Checking to see if the tree structure is the same. 1m nodes from startpos with `-t 1 --smart-pruning-factor=0.0`. Result: differs by a few tens of nodes, which is probably no big deal considering we have 100m nodes.

policy optimized:
```
info string d2d3  (288 ) N:    1879 (+ 0) (P:  2.76%) (WL: -0.00463) (D:  0.322) (M:  5.6) (Q: -0.00463) (U: 0.19765) (S:  0.19302) (V:  -.----) 
info string b2b3  (230 ) N:    2776 (+ 0) (P:  3.28%) (WL:  0.01999) (D:  0.310) (M:  5.3) (Q:  0.01999) (U: 0.15914) (S:  0.17913) (V:  -.----) 
info string b1c3  (36  ) N:    3722 (+ 0) (P:  3.81%) (WL:  0.03385) (D:  0.314) (M:  8.3) (Q:  0.03385) (U: 0.13775) (S:  0.17159) (V:  -.----) 
info string c2c3  (259 ) N:    4847 (+ 0) (P:  4.51%) (WL:  0.04186) (D:  0.320) (M:  6.1) (Q:  0.04186) (U: 0.12528) (S:  0.16714) (V:  -.----) 
info string e2e3  (317 ) N:    8239 (+ 0) (P:  5.57%) (WL:  0.06386) (D:  0.330) (M:  5.5) (Q:  0.06386) (U: 0.09106) (S:  0.15492) (V:  -.----) 
info string c2c4  (264 ) N:    9507 (+ 0) (P:  5.83%) (WL:  0.06931) (D:  0.321) (M:  7.7) (Q:  0.06931) (U: 0.08259) (S:  0.15190) (V:  -.----) 
info string g2g3  (374 ) N:   11008 (+ 0) (P:  6.53%) (WL:  0.07104) (D:  0.325) (M:  6.8) (Q:  0.07104) (U: 0.07984) (S:  0.15088) (V:  -.----) 
info string g1f3  (159 ) N:   16387 (+ 0) (P:  7.92%) (WL:  0.08049) (D:  0.330) (M:  7.1) (Q:  0.08049) (U: 0.06504) (S:  0.14553) (V:  -.----) 
info string d2d4  (293 ) N:   20411 (+ 0) (P:  7.82%) (WL:  0.08921) (D:  0.329) (M:  7.8) (Q:  0.08921) (U: 0.05160) (S:  0.14080) (V:  -.----) 
info string e2e4  (322 ) N:  913018 (+ 0) (P: 33.23%) (WL:  0.11473) (D:  0.329) (M: 15.9) (Q:  0.11473) (U: 0.00490) (S:  0.11963) (V:  -.----) 
```

c4c92f6:
```
info string d2d3  (288 ) N:    1883 (+ 0) (P:  2.76%) (WL: -0.00455) (D:  0.322) (M:  5.6) (Q: -0.00455) (U: 0.19732) (S:  0.19277) (V:  -.----) 
info string b2b3  (230 ) N:    2777 (+ 0) (P:  3.28%) (WL:  0.02004) (D:  0.310) (M:  5.3) (Q:  0.02004) (U: 0.15912) (S:  0.17916) (V:  -.----) 
info string b1c3  (36  ) N:    3725 (+ 0) (P:  3.81%) (WL:  0.03385) (D:  0.314) (M:  8.3) (Q:  0.03385) (U: 0.13761) (S:  0.17146) (V:  -.----) 
info string c2c3  (259 ) N:    4848 (+ 0) (P:  4.51%) (WL:  0.04179) (D:  0.320) (M:  6.1) (Q:  0.04179) (U: 0.12528) (S:  0.16707) (V:  -.----) 
info string e2e3  (317 ) N:    8238 (+ 0) (P:  5.57%) (WL:  0.06377) (D:  0.330) (M:  5.5) (Q:  0.06377) (U: 0.09109) (S:  0.15486) (V:  -.----) 
info string c2c4  (264 ) N:    9509 (+ 0) (P:  5.83%) (WL:  0.06920) (D:  0.321) (M:  7.7) (Q:  0.06920) (U: 0.08261) (S:  0.15181) (V:  -.----) 
info string g2g3  (374 ) N:   11050 (+ 0) (P:  6.53%) (WL:  0.07111) (D:  0.325) (M:  6.8) (Q:  0.07111) (U: 0.07955) (S:  0.15066) (V:  -.----) 
info string g1f3  (159 ) N:   16418 (+ 0) (P:  7.92%) (WL:  0.08044) (D:  0.330) (M:  7.1) (Q:  0.08044) (U: 0.06493) (S:  0.14537) (V:  -.----) 
info string d2d4  (293 ) N:   20491 (+ 0) (P:  7.82%) (WL:  0.08917) (D:  0.329) (M:  7.8) (Q:  0.08917) (U: 0.05141) (S:  0.14058) (V:  -.----) 
info string e2e4  (322 ) N:  913134 (+ 0) (P: 33.23%) (WL:  0.11456) (D:  0.329) (M: 15.9) (Q:  0.11456) (U: 0.00490) (S:  0.11946) (V:  -.----) 
```

**Comment by farmersrice** [comment]
Fixed the corner case Tilps mentioned, new bench:
```
===========================
Total time (ms) : 340447
Nodes searched  : 70244321
Nodes/second    : 206329
```

The 1m nodes test posted above gives the exact same results it did before.

File: src/mcts/search.cc (hunk 1 above)
Comment (farmersrice) [reply src/mcts/search.cc:1453]: Thanks, it is fixed.

File: src/mcts/search.h [hunk 2]
```diff
@@ -328,6 +328,12 @@ class SearchWorker {
   const bool moves_left_support_;
   IterationStats iteration_stats_;
   StoppersHints latest_time_manager_hints_;
+  // Intermediate array to store values when processing policy.
+  // According to a lichess developer post:
+  // https://lichess.org/blog/Wqa7GiAAAOIpBLoY/
+  //     developer-update-275-improved-game-compression
+  // there are never more than 256 valid legal moves in any legal position.
+  float intermediate_[256];
```
Comment (mooskagh) [inline src/mcts/search.h:336]: No need to have it as class member, it's fine to have this on stack inside function.
Without initializer (`= {}`), it's not taking any time to initialize (as it will be left uninitalized).

Even better, use `std::array<float, 256>;` instead. It still won't be initialized.

File: src/mcts/search.cc [hunk 3]
```diff
@@ -1444,16 +1445,16 @@ void SearchWorker::FetchSingleNodeResult(NodeToProcess* node_to_process,
...
     const float scale = 1.0f / total;
-    for (auto edge : node->Edges()) edge.edge()->SetP(edge.GetP() * scale);
+    for (auto edge : node->Edges()) edge.edge()->SetP(intermediate_[counter++] * scale);
+  } else {
+    for (auto edge : node->Edges()) edge.edge()->SetP(intermediate_[counter++]);
   }
```
Comment (Mardak) [inline src/mcts/search.cc:1458]: nit: I don't know if my suggestion is acceptable style, but as mentioned earlier, the expected common case is `total > 0` so slowing down the exceptional case with "unnecessary" multiply might be acceptable…
```c
const auto scale = total > 0.0f ? 1.0f / total : 1.0f;
for (auto edge : node->Edges()) edge.edge()->SetP(intermediate[counter++] * scale);
```

**Comment by farmersrice** [comment]
Match: sf 11 vs. lc0, lc0 policy optimized (pr 1278)
Options:
lc0: all defaults
stockfish: 12 threads, 256mb hash
Time control: 4+0.04 for sf, 5+0.05 for lc0 (~ 25 kn / move for lc0)
Book: Chad 8 ply 10k
Adjudication: -draw movenumber=40 movecount=4 score=8 -resign movecount=4 score=500 
Speed: stockfish 20m nps, lc0 110k nps
Network: last t59
software: cutechess-cli
Results:
```   # PLAYER                  :  RATING  ERROR  POINTS  PLAYED   (%)  CFS(%)     W     D     L  D(%)
   1 sf                      :     0.0   ----  5199.0   10000    52     100  2419  5560  2021    56
   2 lc0_policy_optimized    :   -13.5    6.5  2406.0    5000    48      63  1021  2770  1209    55
   3 lc0_c4c92f6             :   -15.0    6.5  2395.0    5000    48     ---  1000  2790  1210    56

White advantage = 50.34 +/- 2.32
Draw rate (equal opponents) = 57.04 % +/- 0.51
```
Comments:
After 10000 games, no apparent regression for the optimized policy. (lc0_policy_optimized is the binary from right after Tilps's issue was fixed.)

File: src/mcts/search.h (hunk 2 above)
Comment (farmersrice) [reply src/mcts/search.h:336]: Looks good to me:
===========================
Total time (ms) : 340519
Nodes searched  : 70133528
Nodes/second    : 205960

File: src/mcts/search.cc (hunk 3 above)
Comment (farmersrice) [reply src/mcts/search.cc:1458]: Thanks, this is a lot cleaner.
```
===========================
Total time (ms) : 340545
Nodes searched  : 70147040
Nodes/second    : 205984
```

File: src/mcts/search.cc [hunk 4]
```diff
@@ -1435,6 +1435,12 @@ void SearchWorker::FetchSingleNodeResult(NodeToProcess* node_to_process,
                                 edge.GetMove().as_nn_index(
                                     node_to_process->probability_transform)));
   }
+  // Intermediate array to store values when processing policy.
+  // According to a lichess developer post:
```
Comment (mooskagh) [inline src/mcts/search.cc:1439]: I think such a long comment adds more noise than usefulness, especially as it's not related to the function purpose. I'd only leave the last line of it (or combined the first and the list line if they could fit one line together).

File: src/mcts/search.cc [hunk 5]
```diff
@@ -1435,6 +1435,12 @@ void SearchWorker::FetchSingleNodeResult(NodeToProcess* node_to_process,
...
+  // https://lichess.org/blog/Wqa7GiAAAOIpBLoY/
+  //     developer-update-275-improved-game-compression
+  // there are never more than 256 valid legal moves in any legal position.
+  float intermediate[256];
```
Comment (mooskagh) [inline src/mcts/search.cc:1443]: I suggest to use std::array<> instead, it basically has no drawbacks to legacy arrays and only benefits.

Comment (farmersrice) [reply src/mcts/search.cc]: Thanks, it is changed now.
```
===========================
Total time (ms) : 340358
Nodes searched  : 70115712
Nodes/second    : 206005
```

Comment (Mardak) [reply src/mcts/search.cc]: How are you testing? I get a compiler error with clang:
```
FAILED: lc0@exe/src_mcts_search.cc.o 

../../src/mcts/search.cc:1440:26: error: implicit instantiation of undefined template 'std::__1::array<float, 256>'
  std::array<float, 256> intermediate;
                         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tuple:219:64: note: template is declared here
template <class _Tp, size_t _Size> struct _LIBCPP_TEMPLATE_VIS array;
                                                               ^
1 error generated.
```

Comment (farmersrice) [reply src/mcts/search.cc]: I'm using clang too, works for me just fine.
```
C++ compiler for the host machine: clang++-6.0 (clang 6.0.1-11 "clang version 6.0.1-11 (tags/RELEASE_601/final)")
C++ linker for the host machine: clang++-6.0 ld.bfd 2.33
```

File: src/mcts/search.cc [hunk 6]
```diff
@@ -35,6 +35,7 @@
 #include <iterator>
 #include <sstream>
 #include <thread>
+#include <array>
```
Comment (Mardak) [inline src/mcts/search.cc:38]: clang-format keeps these includes sorted

File: src/mcts/search.cc [hunk 7]
```diff
@@ -1444,16 +1449,14 @@ void SearchWorker::FetchSingleNodeResult(NodeToProcess* node_to_process,
     // Note that we want to calculate (exp(p-max_p))^(1/T) = exp((p-max_p)/T).
     p = FastExp((p - max_p) / params_.GetPolicySoftmaxTemp());
 
-    // Note that p now lies in [0, 1], so it is safe to store it in compressed
-    // format. Normalization happens later.
-    edge.edge()->SetP(p);
-    // Edge::SetP does some rounding, so only add to the total after rounding.
-    total += edge.edge()->GetP();
+    intermediate[counter++] = p;
+    total += p;
```
Comment (Mardak) [inline src/mcts/search.cc:1453]: Looks like this resulted in some accuracy improvements for fp16 in addition to the speed optimization. @fohristiwhirl noticed:
```
9.02, 9.01, 7.62, 6.33, 5.82 fp16 0.25.1
9.01, 9.00, 7.62, 6.32, 5.85 fp16 master
9.00, 8.97, 7.59, 6.32, 5.86 my opencl/blas
```
These are the top 5 policies using sv384x30-swa-3010000 from
```
position startpos moves d2d4 g8f6 g1f3 e7e6 c1f4 d7d5 e2e3 c7c5 b1d2 d8b6 a1b1 c5d4 e3d4 c8d7 c2c3 d7b5 d1b3 b5f1 b3b6 a7b6 h1f1 b8c6 a2a3 f6h5 f4g3 f7f6 e1d1 g7g5 d1c2 a8c8 d2b3 h5g7 h2h3 g7f5 g3h2 h7h5 b1e1 e8f7 c2b1 g5g4 f3d2 f8h6 e1e2 h8g8 h3g4 g8g4 g2g3 c6a5 f1e1 c8e8 b1c2 a5c6 d2f3 h6f8 b3c1 f8d6 e1d1 g4e4 d1e1 e4e2 e1e2 e6e5 d4e5 f6e5 c2d1 e5e4 f3d2 c6e5 d1c2 d6c5 a3a4
```
